controllers/neutronapi_controller.go

```diff
@@ -89,327 +98,467 @@ func (r *NeutronAPIReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error)
-					Image: instance.Spec.ContainerImage,
-					// TODO - tripleo healthcheck script expects vhost config at /etc/httpd/conf.d/10-nova_api_wsgi.conf
-					//ReadinessProbe: readinessProbe.GetProbe(),
-					//LivenessProbe:  livenessProbe.GetProbe(),
-					Env:          envs,
-					VolumeMounts: volumeMounts,
-				},
-			},
-		}
+	// Handle service upgrade
+	ctrlResult, err = r.reconcileUpgrade(ctx, instance, helper)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
 
-		initContainerDetails := common.InitContainer{
-			ContainerImage: instance.Spec.ContainerImage,
-			DatabaseHost:   instance.Spec.DatabaseHostname,
-			NeutronSecret:  instance.Spec.NeutronSecret,
-			NovaSecret:     instance.Spec.NovaSecret,
-			VolumeMounts:   initVolumeMounts,
-		}
-		deployment.Spec.Template.Spec.InitContainers = common.GetInitContainer(initContainerDetails)
+	// Define a new Deployment object
+	depl := common.NewDeployment(
+		neutronapi.Deployment(instance, inputHash, serviceLabels),
+		5,
+	)
 
-		err := controllerutil.SetControllerReference(instance, deployment, r.Scheme)
-		if err != nil {
-			return err
-		}
+	ctrlResult, err = depl.CreateOrPatch(ctx, helper)
+	if err != nil {
+		return ctrlResult, err
+	} else if (ctrlResult != ctrl.Result{}) {
+		return ctrlResult, nil
+	}
+	// create Deployment - end
+
+	r.Log.Info("Reconciled Service successfully")
+	return ctrl.Result{}, nil
+}
+
+//
+// generateServiceConfigMaps - create create configmaps which hold scripts and service configuration
+// TODO add DefaultConfigOverwrite
+//
+func (r *NeutronAPIReconciler) generateServiceConfigMaps(
+	ctx context.Context,
+	h *helper.Helper,
+	instance *neutronv1beta1.NeutronAPI,
+	envVars *map[string]common.EnvSetter,
+) error {
+	// Create/update configmaps from templates
+	cmLabels := common.GetLabels(instance, common.GetGroupLabel(neutronapi.ServiceName), map[string]string{})
+
+	// customData hold any customization for the service.
+	// custom.conf is going to /etc/<service>/<service>.conf.d
+	// all other files get placed into /etc/<service> to allow overwrite of e.g. logging.conf or policy.json
+	// TODO: make sure custom.conf can not be overwritten
+	customData := map[string]string{common.CustomServiceConfigFileName: instance.Spec.CustomServiceConfig}
+	for key, data := range instance.Spec.DefaultConfigOverwrite {
+		customData[key] = data
+	}
+
+	keystoneAPI, err := keystone.GetKeystoneAPI(ctx, h, instance.Namespace, map[string]string{})
+	if err != nil {
+		return err
+	}
+	authURL, err := keystoneAPI.GetEndpoint(common.EndpointPublic)
+	if err != nil {
+		return err
+	}
+	templateParameters := make(map[string]interface{})
+	templateParameters["ServiceUser"] = instance.Spec.ServiceUser
+	templateParameters["KeystonePublicURL"] = authURL
 
+	ovnConnection, _, err := common.GetConfigMapAndHashWithName(ctx, h, instance.Spec.OVNConnectionConfigMap, instance.Namespace)
```

Comment:
I think you're already getting this in `reconcileNormal`, so you could pass it to this function to avoid querying `etcd` again.